File: src/ls/_run_ls_params.R
# This file has been modified to work with 
# our automated system for the biochronicity project.
# 2012 01 30
# INPUTS:
#	FOR FILES & DIRECTORIES	
#	1.	data_file: (string) the data file to run through the alg.
#		File should be tab delimited, no quotes. 
#		The first row contains time labels (float). 
#		The first column contains the IDs (string). 
#		Position 1,1 usually has some text (like "probe"). 
#		The cell values are numeric (float).
#	2.	results_dir: (string) path to directory where files will be saved.
#	3.	results_name: (string) first part of file name for all saved files.
#	FOR ALGORITHM
#	4.	per_min: (float|int) the minimum period for range of periods to examine.
#	5.	per_max: (float|int) the maximum period for range of periods to examine.
#	6.	test_freq:	(float) number of test frequencies to scan,
#		test_freq * #timepoints, so this is multiplier. Suggested 2 or 4.
#	7.	unit_type: (string) unit for timepoints, e.g. "min", "hr". etc.
#
# SAVES:
#	1.	<results_dir>/<results_name>_Dominant.tsv
#		for each record, lists the information for the "best" period.
#		tab delimited, no quotes.
#	2.	<results_dir>/<results_name>_pValues.tsv
#		for each record, reports the adjusted pval for each period tested.
#		tab delimited, no quotes.


# Example of processing time series set
# efg, Stowers Institute for Medical Research, July 2005

source("../src/ls/ProcessLombScargleDataset.R")


args = commandArgs(trailingOnly = TRUE)
data_file = args[1]
results_dir = args[2]
results_name = args[3] 
per_min = as.numeric(args[4])
per_max = as.numeric(args[5])
test_freq = as.numeric(args[6])
unit_type = args[7]


# ====================================================================
# 1. Read data into Data Frame called RawData
filename <- data_file

# (use as.is=TRUE if IDs in column 1 are NOT unique
RawData <- read.table(file=filename, header=FALSE, sep="\t", quote="", comment.char = "#", stringsAsFactors = FALSE)
#print("Sample of:")
#print(filename)
#print(RawData[1:10,])

# ====================================================================
# 2. Define data elements needed for Lomb-Scargle Analysis

# 2.1  First column has IDs
ID <- as.character(RawData[2:nrow(RawData),1])
#cat("FIRST ID: ", ID[1], "\n")
#cat("LAST ID:  ", ID[nrow(RawData)-1], "\n")

# 2.2  Define time points for time series (need not be uniform)
# (do not use t for time here since it is reserved in R for transpose)
Time <- as.numeric(RawData[1, 2:ncol(RawData)])
#cat("TIMEPOINTS: ", Time, "\n")

# 2.3  Remaining columns have expression time series
Expression <- data.matrix(RawData[2:nrow(RawData),2:ncol(RawData)])
#cat("FIRST DATA ROW: ", Expression[1,], "\n")
#cat("LAST DATA ROW:  ", Expression[nrow(RawData)-1,], "\n")


# Check for consistency between Expression and Time
N <- ncol(Expression)  # number of time points in series
stopifnot(length(Time) == N)

# Define global value used by Lomb-Scargle to label plots
unit <<- unit_type

# 2.4  Define number of test frequencies:  normally 2N or 4N
M <- N * test_freq
print(M)

# Frequencies usually easier to define in terms of 1/Period.
# In this case since "dominant" frequency is expected to correspond
# to 24-hour period, lets search frequencies corresponding to
# periods from 6 hours to 48 hours.
MinFrequency <- 1 / per_max
MaxFrequency <- 1 / per_min

# See if MaxFrequency is perhaps too high
if (MaxFrequency > 1/(2*mean(diff(Time))))
{
  cat("MaxFrequency may be above Nyquist limit.\n")
}

TestFrequencies <- MinFrequency +
                  (MaxFrequency - MinFrequency) * (0:(M-1) / (M-1))

# 2.5 Define relative path for results and make sure directories exist

SLASH <- .Platform$file.sep
if (! file.exists(results_dir) )
{
  dir.create(results_dir)

  # only needed if SavePlots=TRUE (see below)
  # The "chart" directory will contain PDFs of the Lomb-Scargle analysis
  dir.create(paste(results_dir, SLASH, "charts", sep=""))
}



# ====================================================================
# 3. Process the time series
ProcessExpressionDataset(results_dir, results_name,
                         Time, ID, Expression,
                         TestFrequencies,
                         SavePlots=FALSE)

File: src/ls/_run_ls_params.py
#!/Library/Frameworks/Python.framework/Python
#
# _run_ls_params.py
# BY:	Anastasia Deckard anastasia<dot>deckard<at>duke<dot>edu
# DATE:	2011 01 14
# DESCRIPTION
#	for running algorithm; creates dir and file names, collects params,
#	makes dir in results folder for alg if necessary, makes dir in alg folder
#	for run, saves parameter file, and executes alg with params.
# INPUTS
#	FOR FILES & DIRECTORIES
#	1.	data_file: the data file to run through the alg
#	2.	results_dir: general results directory for repository.
#		will add alg abbr to the path for you.
#	FOR ALGORITHM (see _run_ls_params.R for detailed descriptions)
#	3.	per_min: the minimum period to examine
#	4.	per_max: the maximum period to examine
#	5.	test_freq:	number of test frequencies to scan,
#		test_freq * #timepoints, so this is multiplier.
# ACTIONS
#	1. makes a run name of the form <data_name>_ls_<params>
#	2. makes a directory <results_dir>/ls
#	3. makes a directory <results_dir>/ls/<run_name>
#	4. writes parameters to file <results_dir>/ls/<run_name>/<run_name>_params.txt
#	5. calls _run_ls_params.R with arguments
#	6. saves log to file <results_dir>/ls/<run_name>/<run_name>_log.txt


import os, sys, datetime


#THE ALGORITHM"S ABBR FOR NAMING FILES AND DIRECTORIES
alg = "ls"

#PARAMETERS FOR RUNNING ALGORITHM
#default params
test_freq = 4
unit_type = "minutes"

#user specified params
if (len(sys.argv) == 5):	#using default params
	data_file = sys.argv[1]
	results_dir = sys.argv[2]
	per_min = sys.argv[3]
	per_max = sys.argv[4]
	print("using arguments from command line:")
elif (len(sys.argv) == 7):	#using all speficied params
	data_file = sys.argv[1]
	results_dir = sys.argv[2]
	per_min = sys.argv[3]
	per_max = sys.argv[4]
	test_freq = sys.argv[5]
	unit_type = sys.argv[6]
	print("using arguments from command line:")
else:
	print("Please provide 4 arguments (defaults for other params will be used):")
	print("data_file results_dir per_min per_max")
	print("Please provide 6 arguments:")
	print("data_file results_dir per_min per_max test_freq unit_type")
	exit()

print(data_file, results_dir, per_min, per_max, test_freq, unit_type)



#SETUP NAMES FOR DIRECTORIES AND FILES
data_name = os.path.splitext(os.path.basename(data_file))[0]
run_name = "%s_%s_p%s-%sf%s" %(data_name, alg, per_min, per_max, test_freq)
resrun_dir = os.path.join(results_dir, run_name)
print("run name: ", run_name)
print("run results dir: ", resrun_dir)

# resalg_dir = results_dir #os.path.join(results_dir, alg)
# resrun_dir = os.path.join(resalg_dir, run_name)
# print("run name: ", run_name)
# print("alg results dir: ", resalg_dir)
# print("run results dir: ", resrun_dir)


#MAKE RESULTS DIRECTORY
if not os.path.exists(resrun_dir):
	os.mkdir(resrun_dir)
	print("made: ", resrun_dir)


#WRITE PARAMETERS FILE TO OUT DIRECTORY
param_path = resrun_dir + "/" + run_name + "_params.txt"
param_file = open(param_path,"w")
param_file.write("data_file:\t" + data_file + "\n")
param_file.write("per_min:\t" + str(per_min) + "\n")
param_file.write("per_max:\t" + str(per_max) + "\n")
param_file.write("test_freq:\t" + str(test_freq) + "\n")
param_file.write("unit:\t" + unit_type + "\n")
param_file.close()

#MAKE COMMAND & RUN ALGORITHM, TIME IT
consout_path = resrun_dir + "/" + run_name + "_consout.txt"

cmd1 = "Rscript ../src/ls/_run_ls_params.R %s %s %s %s %s %s %s | tee %s" %(os.path.abspath(data_file), os.path.abspath(resrun_dir), run_name, per_min, per_max, test_freq, unit_type, os.path.abspath(consout_path))
print(cmd1)

time_beg = datetime.datetime.now()

os.system(cmd1)

time_end = datetime.datetime.now()
diff = time_end - time_beg

print("finished, with time: ", str(diff))

#WRITE OUT LOG FILE WITH TIME INFO
log_path = resrun_dir + "/" + run_name + "_log.txt"
log_file = open(log_path,"w")
log_file.write(cmd1 + "\n\n")
log_file.write("start:\t" + time_beg.strftime("%Y-%m-%d %H:%M:%S") + "\n")
log_file.write("end:\t" + time_end.strftime("%Y-%m-%d %H:%M:%S") + "\n")
log_file.write("time:\t" + str(diff) + "\n")
log_file.write("time (S):\t" + str(diff.total_seconds()) + "\n")
log_file.write("time (M):\t" + str(diff.total_seconds()/60) + "\n")
log_file.write("time (H):\t" + str(diff.total_seconds()/(60*60)) + "\n")
log_file.close()
